src/drivers/irlock/irlock.cpp

```diff
@@ -0,0 +1,410 @@
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/*
+ * @file irlock.cpp
+ *
+ * Driver for an IR-Lock and Pixy vision sensor connected via I2C.
+ *
+ */
+
+#include <fcntl.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/types.h>
+
+#include <drivers/boards/px4fmu-v2/board_config.h>
+#include <drivers/device/i2c.h>
+#include <drivers/device/ringbuffer.h>
+#include <drivers/drv_irlock.h>
+#include <drivers/drv_hrt.h>
+
+#include <nuttx/clock.h>
+#include <nuttx/wqueue.h>
+#include <systemlib/err.h>
+
+/* Configuration Constants */
+#define IRLOCK_I2C_BUS			PX4_I2C_BUS_EXPANSION
+#define IRLOCK_I2C_ADDRESS		0x65 //* 7-bit address (non shifted)
+#define IRLOCK_CONVERSION_INTERVAL_US	20000U /* us = 20ms = 50Hz */
+
+#define IRLOCK_SYNC			0xAA55
+#define IRLOCK_RESYNC		0x5500
+#define IRLOCK_ADJUST		0xAA
+
+#ifndef CONFIG_SCHED_WORKQUEUE
+# error This requires CONFIG_SCHED_WORKQUEUE.
+#endif
+
+class IRLOCK : public device::I2C
+{
+public:
+	IRLOCK(int bus = IRLOCK_I2C_BUS, int address = IRLOCK_I2C_ADDRESS);
+	virtual ~IRLOCK();
+
+	virtual int init();
+	virtual int probe();
+	virtual int info();
+
+	virtual ssize_t read(struct file *filp, char *buffer, size_t buflen);
+
+private:
+
+	/* start periodic reads from sensor */
+	void start();
+
+	/* stop periodic reads from sensor */
+	void stop();
+
+	/* static function that is called by worker queue, arg will be pointer to instance of this class */
+	static void		cycle_trampoline(void *arg);
+
+	/* read from device and schedule next read */
+	void			cycle(); \
+
+	/* low level communication with sensor */
+	int read_device();
+	bool sync_device();
+	int read_device_word(uint16_t *word);
+	int read_device_block(struct irlock_s *block);
+
+	/* internal variables */
+	RingBuffer *_reports;
+	bool _sensor_ok;
+	work_s _work;
+};
+
+/* global pointer for single IRLOCK sensor */
+namespace
+{
+IRLOCK *g_irlock = nullptr;
+}
+
+void irlock_usage();
+
+extern "C" __EXPORT int irlock_main(int argc, char *argv[]);
+
+/* constructor */
+IRLOCK::IRLOCK(int bus, int address) :
+	I2C("irlock", IRLOCK0_DEVICE_PATH, bus, address, 400000),
+	_reports(nullptr),
+	_sensor_ok(false)
+{
+	memset(&_work, 0, sizeof(_work));
+}
+
+/* destructor */
+IRLOCK::~IRLOCK()
+{
+	stop();
+
+	/* clear reports queue */
+	if (_reports != nullptr) {
+		delete _reports;
+	}
+}
+
+/* initialise driver to communicate with sensor */
+int IRLOCK::init()
+{
+	/* initialise I2C bus */
+	int ret = I2C::init();
+
+	if (ret != OK) {
+		return ret;
+	}
+
+	/* allocate buffer storing values read from sensor */
+	_reports = new RingBuffer(135, sizeof(struct irlock_s));
```

Comment:
Is the number 135 here based on a particular metric? It will consume RAM, so it would be good to know that this buffer depth is required.